grin/src/miner.rs

```diff
@@ -527,22 +621,182 @@ impl Miner {
+			miner_config.stratum_server_addr.unwrap()
+		);
+
+		// Listen for miner connection
+		for stream in listener.incoming() {
+			stream_err = false;
+			match stream {
+				Err(e) => {
+					// connection failed
+					error!(
+						LOGGER,
+						"(Server ID: {}) Error accepting stratum connection: {:?}",
+						self.debug_output_id,
+						e
+					);
+				}
+				Ok(mut stream) => {
+					info!(LOGGER, "New connection: {}", stream.peer_addr().unwrap());
+					stream
+						.set_nonblocking(true)
+						.expect("set_nonblocking call failed");
+					let mut current_hash = self.chain.head().unwrap().prev_block_h;
+
+					loop {
+						trace!(LOGGER, "key_id: {:?}", key_id);
+
+						// Abort connection on error
+						if stream_err == true {
+							warn!(
+								LOGGER,
+								"(Server ID: {}) Resetting stratum server connection",
+								self.debug_output_id
+							);
+							match stream.shutdown(Shutdown::Both) {
+								_ => {}
+							}
+							break;
+						}
+
+						// get the latest chain state and build a block on top of it
+						let head = self.chain.head_header().unwrap();
+						let mut latest_hash = self.chain.head().unwrap().last_block_h;
+
+						// Build a new block to mine
+						if current_hash != latest_hash || time::get_time().sec >= deadline {
+							// There is a new block on the chain
+							// OR we are rebuilding the current one to include new transactions
+							if current_hash != latest_hash {
+								// A brand new block, so we will generate a new key_id
+								key_id = None;
+							}
+
+							let mut result = self.build_block(&head, key_id.clone());
+							while let Err(e) = result {
+								match e {
+									self::Error::Chain(chain::Error::DuplicateCommitment(_)) => {
+										debug!(LOGGER,
+										       "Duplicate commit for potential coinbase detected. Trying next derivation.");
+									}
+									ae => {
+										warn!(
+											LOGGER,
+											"Error building new block: {:?}. Retrying.", ae
+										);
+									}
+								}
+								thread::sleep(Duration::from_millis(100));
+								result = self.build_block(&head, key_id.clone());
+							} // error building block
+
+							let (new_block, block_fees) = result.unwrap();
+							b = new_block;
+							key_id = block_fees.key_id();
+							current_hash = latest_hash;
+							// set a new deadline for rebuilding with fresh transactions
+							deadline = time::get_time().sec + attempt_time_per_block as i64;
+
+							debug!(
+								LOGGER,
+								"(Server ID: {}) sending block {} to stratum client",
+								self.debug_output_id,
+								b.header.height
+							);
+
+							// Push the new block "b" to the connected client
+							self.stratum_send_block(&mut stream, &b, &mut stream_err);
+						}
+
+						// Get a solved block header - if any are waiting
+						// Here, we are expecting real solutions at the full difficulty.
+						match self.stratum_get_solution(&mut stream, &mut stream_err) {
```

Comment:
I know it's WIP, but we definitely need to be getting back solutions of any difficulty, not just the full, so that a client can prove work.